text/0000-fields-in-traits.md

```diff
@@ -0,0 +1,800 @@
+the trait and define a field with that name. Sometimes it may be
+necessary or desirable to specify the trait explicitly. For those
+cases, we introduce a fully qualified field notation which looks like
+`x.<Trait<U,V>::f>`.
+
+This is comparable to the associated item notation `<T as
+Trait<U,V>>::foo`, but with some differences. First, the `Self` type
+is implied by the type of the left-hand side (`x`, in our
+example). Second, it is part of an lvalue expression, rather than
+being a path to an item.
+
+This is obviously a point where one might bikeshed on the best syntax.
+Here are some alternatives that we considered and rejected:
+
+- `x.Trait::f` more closely resembles C++ and Java and could work, but
+  it would have to be `x.Trait::<U,V>::f` if you wish to specify the trait's
+  type parameters.
+- `<x as Trait<U,V>>.f` doesn't work because we'd need a cover grammar for types
+  and expressions, since you cannot do not know whether `x` is a type or
+  expression until you see the `.`.
+- `x::<as Trait<U,V>>.f` seems random
+
+#### Leaving room for method calls
+
+It has not escaped our notice that this same syntax could be used for
+method calls. This RFC does not propose that we permit that, but we do
+restrict the grammar of call expressions to disallow a fully qualified
+field path on the left-hand-side of `()`. In other words, just as
+calling a closure located at `a.b.c` must be written `(a.b.c)()` (so
+as to avoid ambiguity with calling a method `c` on the path `a.b`), so
+must you wrote `(a.b.<Trait::c>)()`; `a.b.<Trait::c>()` will not
+parse.
+
+### Field access via traits
+
+This section covers the main rules regarding integrating trait fields
+into the borrow checker and other safety analyses.
+
+#### When are two fields disjoint?
+
+Today, if I access two fields of a struct like `base.a` and `base.b`,
+those two paths are known to be disjoint by the borrow checker. In the
+case where one or both of those fields is defined by a trait, we must
+now generalize this notion and define the criteria when two field
+names `a` and `b` are considered disjoint from one another (we assume
+both are being accessed from the same path `base`):
+
+- Both fields defined in the same struct: disjoint
+- Field `a` is defined in a trait `A`, field `b` is defined in a trait `B`:
+  disjoint `A` is a supertrait of `B` or vice versa
+  (not that every trait is its own supertrait)
+- Otherwise: potentially overlapping   
+
+#### Moves are not allowed
+
+For soundness reasons, we disallow moving individual fields out of
+values that have a defined destructor. When writing generic code, or
+when working with objects, one can never know whether a value has a
+destructor or not. Therefore, we disallow moving out of a trait field
+if the owner of that field may have a destructor. In practice that means
+that moving out from a field owned by one of these sorts of types is
+disallowed:
+
+- trait objects (e.g., `Write`);
+- generic type parameters (`T`);
```

Comment:
And what about pattern destructuring the type, binding only to the fields the trait has defined? Something like this?

```rust
let T { field1, field2, .. } = self;
```